Makefile

```diff
@@ -1,51 +1,131 @@
-RETOOL=$(CURDIR)/_tools/bin/retool
-PATH := ${PWD}/bin:${PWD}/ENV/bin:${PATH}
-.DEFAULT_GOAL := all
+DEP_VERSION := 0.5.0
+PROTOC_VERSION := 3.7.0
+ERRCHECK_VERSION := 1.2.0
+PROTOC_GEN_GO_VERSION := 1.3.0
+PROTOC_GEN_GOFAST_VERSION := 1.2.1
 
-all: setup test_all
+SHELL := /bin/bash -o pipefail
+UNAME_OS := $(shell uname -s)
+UNAME_ARCH := $(shell uname -m)
 
-.PHONY: test test_all test_core test_clients test_go_client test_python_client generate
+TMP_BASE := .tmp
+TMP := $(TMP_BASE)/$(UNAME_OS)/$(UNAME_ARCH)
+TMP_BIN = $(TMP)/bin
+TMP_VENV = $(TMP)/venv
+TMP_VERSIONS := $(TMP)/versions
 
-# Phony commands:
-generate:
-	PATH=$(CURDIR)/_tools/bin:$(PATH) GOBIN="${PWD}/bin" go install -v ./protoc-gen-...
-	$(RETOOL) do go generate ./...
+$(TMP_BIN):
+	@mkdir -p $(TMP_BIN)
 
-test_all: setup test_core test_clients
+$(TMP_VENV):
+	@mkdir -p $(dir $(TMP_VENV))
+	virtualenv $(TMP_VENV)
```

Comment:
Virtualenv needs to be installed, or called out as a dependency (probably the latter).